packages/salesforcedx-vscode-core/src/orgBrowser/metadataOutlineProvider.ts

```diff
@@ -24,33 +37,144 @@ export class MetadataOutlineProvider
+      return Promise.resolve(element.children);
+    } else if (element.type === NodeType.MetadataType) {
+      const metadataCmps = await this.getComponents(element);
+      if (isNullOrUndefined(metadataCmps) || metadataCmps.length === 0) {
+        const emptyDefault = new BrowserNode(
+          'No components available',
+          NodeType.EmptyNode
+        );
+        element.children = [emptyDefault];
+      } else {
+        element.children = metadataCmps;
       }
-    } else {
       return Promise.resolve(element.children);
     }
+
+    return Promise.resolve([]);
   }
 
-  public async getDefaultUsernameOrAlias() {
+  public async getTypes(): Promise<BrowserNode[]> {
+    const username = this.defaultOrg!;
+    const typesFolder = await getTypesFolder(username);
+    const typesPath = path.join(typesFolder, 'metadataTypes.json');
+
+    let typesList: string[];
+    if (!fs.existsSync(typesPath)) {
+      try {
+        const result = await forceDescribeMetadata(typesFolder);
+        typesList = buildTypesList(result, undefined);
+      } catch (e) {
+        const errorNode = new BrowserNode(parseErrors(e), NodeType.EmptyNode);
+        return [errorNode];
+      }
+    } else {
+      typesList = buildTypesList(undefined, typesPath);
+    }
+    const nodeList = [];
+    for (const type of typesList) {
+      const typeNode = new BrowserNode(type, NodeType.MetadataType);
+      nodeList.push(typeNode);
+    }
+    return nodeList;
+  }
+
+  public async getComponents(
+    metadataType: BrowserNode
+  ): Promise<BrowserNode[]> {
+    const username = this.defaultOrg!;
+    const componentsPath = await getComponentsPath(
+      metadataType.label!,
+      username
+    );
+
+    let componentsList: string[];
+    if (!fs.existsSync(componentsPath)) {
+      const result = await forceListMetadata(
+        metadataType.label!,
+        username,
+        componentsPath
+      );
+      componentsList = buildComponentsList(
+        metadataType.label!,
+        result,
+        undefined
+      );
+    } else {
+      componentsList = buildComponentsList(
+        metadataType.label!,
+        undefined,
+        componentsPath
+      );
+    }
+    const nodeList = [];
+    for (const component of componentsList) {
+      const cmpNode = new BrowserNode(component, NodeType.MetadataCmp);
+      nodeList.push(cmpNode);
+    }
+    return nodeList;
+  }
+
+  public async getDefaultUsernameOrAlias(): Promise<boolean> {
     if (hasRootWorkspace()) {
-      this.defaultOrg = await OrgAuthInfo.getDefaultUsernameOrAlias(false);
+      const username = await OrgAuthInfo.getDefaultUsernameOrAlias(false);
+      let diff = false;
+      if (username !== this.defaultOrg) {
+        diff = true;
+      }
+      this.defaultOrg = username;
+      return diff;
+    } else {
+      throw new Error('Workspace could not be found.');
     }
   }
 }
+
+export function parseErrors(error: string): string {
+  const e = JSON.parse(error);
```

Comment:
Do we want to change the error message based on the type of error?

Reply:
Yes, what errors are not already covered here ?

Reply:
This was all of them! Was just checking if we wanted these.